espei/paramselect.py

```diff
@@ -422,15 +417,15 @@ def phase_fit(dbf, phase_name, symmetry, datasets, refdata, ridge_alpha, aicc_pe
             check_symbol = Symbol('YS') * Symbol('Z')**degree
             keys_to_remove = []
             for key, value in sorted(parameters.items(), key=str):
-                if key.has(check_symbol):
+                if key.xreplace({check_symbol: 0}) != key:
```

Comment:
In manual testing, I found that this workaround for `has` doesn't work for interactions with temperature terms. I added a test (`test_parameter_generation.test_high_order_interaction_terms_no_spurious_symbols`) to exercise the bug.

The output from that test contains

```python
TRACE:espei.paramselect - Polynomial coefs: [T*VV0000 + T*Z*VV0001, 0.0, 0.0, 0.0, 0, 0, 0, 0, 0, 0]
```

i.e. what's happening is that the `YS*Z` feature doesn't get xreplaced when `degree = 1`, but the `YS` (constant T) feature that becomes V0000 gets replaced. 

Here's some code to reproduce:

```python
# Uncomment one import or the other to test
# from symengine import Symbol
# from sympy import Symbol

key = Symbol("YS") * Symbol("Z")
key_with_T = Symbol("T") * Symbol("YS") * Symbol("Z")
check_symbol = Symbol("YS") * Symbol("Z")
print("check_symbol", check_symbol)
# Sympy and SymEngine both give the same results here:
# When key == check_symbol, the xreplace works; when T is multiplied, the xreplace doesn't replace
print("key", key)
print("key replaced", key.xreplace({check_symbol: 0}))
print("key_with_T", key_with_T)
print("key_with_T replaced", key_with_T.xreplace({check_symbol: 0}))

# Only SymPy works (gives `True`), SymEngine raises with
# TypeError: <class 'symengine.lib.symengine_wrapper.Symbol'> required. Z*YS is of type <class 'symengine.lib.symengine_wrapper.Mul'>
print(key.has(check_symbol))
```

Reply:
I think this is https://github.com/symengine/symengine/issues/1359

Reply:
This function will pass the test you made:

```python
def has_symbol(expr, check_symbol):
    """
    Workaround for SymEngine not supporting Basic.has() with non-Symbol arguments.
    Only works for detecting subsets of multiplication of variables.
    """
    try:
        return expr.has(check_symbol)
    except TypeError:
        for at in expr.atoms(symengine.Mul):
            if set(check_symbol.args).issubset(at.args):
                return True
    return False
```

However, it will fail when dealing with powers of variables in the expression:
```python
has_symbol(Symbol("YS") * (Symbol("Z") ** 2), Symbol("YS") * (Symbol("Z")))
# returns False, but is actually True
````

Reply:
The purpose of this chunk of code is to associate the symbolic representation of a feature `T*YS*Z` to the RK interaction order (`degree`) so it can be entered into the database. We go in reverse order so that the high order terms are matched and removed first. 

This could be a good start as a workaround if we extend it for powers. There might be some other changes we could make to accomplish the same thing

Reply:
This seems to work:
```python
def _poly_degrees(expr):
    poly_dict = {}
    for at in expr.atoms(symengine.Symbol):
        poly_dict[at] = 1
    for at in expr.atoms(symengine.log):
        poly_dict[at] = 1
    for at in expr.atoms(symengine.Pow):
        poly_dict[at.args[0]] = at.args[1]
    return poly_dict

def has_symbol(expr, check_symbol):
    """
    Workaround for SymEngine not supporting Basic.has() with non-Symbol arguments.
    Only works for detecting subsets of multiplication of variables.
    """
    try:
        return expr.has(check_symbol)
    except TypeError:
        expr_poly = _poly_degrees(expr)
        check_poly = _poly_degrees(check_symbol)
        print(expr, check_symbol, expr_poly, check_poly)
        for cs, check_degree in check_poly.items():
            eps = expr_poly.get(cs, 0)
            if check_degree > eps:
                return False
        return True
```